File: picourl/src/main/java/com/commit451/picourl/PicoUrl.java
package com.commit451.picourl;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;
import java.util.concurrent.Callable;

import io.reactivex.Single;
import io.reactivex.SingleSource;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.ResponseBody;
import retrofit2.Response;

/**
 * Shorten your urls, powered by http://tinyurl.com/. Use the {@link Builder}
 * to create an instance.
 */
public class PicoUrl {

    /**
     * Build your {@link PicoUrl} instance
     */
    public static class Builder {
        private String baseUrl;
        private String param;
        private OkHttpClient client;

        /**
         * Create a new Builder
         */
        public Builder() {
        }

        /**
         * Set the base url to use.
         * @param baseUrl the base url
         * @return builder
         */
        public Builder baseUrl(@NonNull String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }

        /**
         * Set the {@link OkHttpClient} PicoUrl will use. Make sure to include any configuration
         * or authentication you might need
         * @param client the client
         * @return builder
         */
        public Builder client(@Nullable OkHttpClient client) {
            this.client = client;
            return this;
        }

        /**
         * Set the query param that will show up in the shortened url. By default, "tinyUrl"
         * @param param the key that will be used for the shortened url
         * @return builder
         */
        public Builder tinyQueryParam(@Nullable String param) {
            this.param = param;
            return this;
        }

        /**
         * Build the {@link PicoUrl} instance
         * @return the built instance
         */
        @NonNull
        public PicoUrl build() {
            if (baseUrl == null) {
                throw new IllegalStateException("You need to specify a base url");
            }
            if (client == null) {
                client = new OkHttpClient();
            }
            if (param == null) {
                param = DEFAULT_QUERY_PARAM_TINY_URL;
            }
            return new PicoUrl(baseUrl, param, client);
        }
    }

    private static final String DEFAULT_QUERY_PARAM_TINY_URL = "tinyUrl";

    private TinyUrl tinyUrl;
    private String baseUrl;
    private String param;
    private OkHttpClient client;

    private PicoUrl(String baseUrl, String param, OkHttpClient client) {
        this.baseUrl = baseUrl;
        this.client = client;
        this.param = param;
        tinyUrl = TinyUrlFactory.create(client);
    }

    /**
     * Generate a shortened url, eg. http://yourbaseurl.com?tinyUrl=asdfhds
     * @param url the url to shorten.
     * @return an observable with the shorted url
     */
    public Single<Uri> generate(@NonNull final Uri url) {
        return Single.defer(new Callable<SingleSource<? extends Uri>>() {
            @Override
            public SingleSource<? extends Uri> call() throws Exception {
                Uri generatedUrl = generateInternal(url);
                return Single.just(generatedUrl);
            }
        });
    }

    /**
     * Parse a shorted url
     * @param url the url that was shorted by the {@link #generate(Uri)} method
     * @return the original url that was passed to the {@link #generate(Uri)} method
     */
    public Single<Uri> parse(@NonNull final Uri url) {
        return Single.defer(new Callable<SingleSource<? extends Uri>>() {
            @Override
            public SingleSource<? extends Uri> call() throws Exception {
                return Single.just(parseInternal(url));
            }
        });
    }

    private Uri generateInternal(final Uri url) throws IOException {
        Response<ResponseBody> tinyUrlResponse = tinyUrl.generateLink(url.toString()).execute();
        String tinyUrl = tinyUrlResponse.body().string();
        //Get the unique string at the end of the tinyurl
        String tinyUrlPath = tinyUrl.split(".com/")[1];
        //Builds our url like so:
        //http://yourdomain.com&tinyurl=asdfwe
        return Uri.parse(baseUrl).buildUpon()
                .appendQueryParameter(param, tinyUrlPath)
                .build();
    }

    private Uri parseInternal(Uri url) throws Exception {
        String tinyUrlPath = url.getQueryParameter(param);
        if (tinyUrlPath == null) {
            throw new IllegalArgumentException("Passed url does not contain a tiny url param");
        }
        final String followUrl = TinyUrl.BASE_URL + tinyUrlPath;
        Request followRequest = new Request.Builder()
                .url(followUrl)
                .build();
        okhttp3.Response response = client.newCall(followRequest).execute();

        String followedUrl = Util.backpedalRedirectsTillYouDie(response, baseUrl);
        if (followedUrl == null) {
            response.body().close();
            //Throw some more meaningful error
            throw new RedirectNotFoundException();
        }
        //We have to do this now, since really we had our urls as params
        final String parsedUrl = Uri.decode(followedUrl);
        response.body().close();
        return Uri.parse(parsedUrl);
    }
}

File: picourl/src/main/kotlin/com/commit451/picourl/PicoUrl.kt
package com.commit451.picourl

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.request
import io.ktor.http.HttpHeaders
import io.ktor.http.URLBuilder
import io.ktor.http.Url
import io.ktor.http.isSuccess
import java.io.Closeable
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets

/**
 * Shorten your URLs, powered by tinyurl.com.
 */
class PicoUrl private constructor(
    private val baseUrl: String,
    private val param: String,
    private val client: HttpClient,
    private val ownsClient: Boolean,
) : Closeable {

    class Builder {
        private var baseUrl: String? = null
        private var param: String? = null
        private var client: HttpClient? = null

        fun baseUrl(baseUrl: String) = apply {
            this.baseUrl = baseUrl
        }

        fun client(client: HttpClient?) = apply {
            this.client = client
        }

        fun tinyQueryParam(param: String?) = apply {
            this.param = param
        }

        fun build(): PicoUrl {
            val requiredBaseUrl = baseUrl ?: error("You need to specify a base url")
            val requiredParam = param ?: DEFAULT_QUERY_PARAM_TINY_URL
            val configuredClient = client ?: picoUrlHttpClient()
            return PicoUrl(
                baseUrl = requiredBaseUrl,
                param = requiredParam,
                client = configuredClient,
                ownsClient = client == null,
            )
        }
    }

    suspend fun generate(url: URI): URI = URI(generate(url.toString()))

    suspend fun generate(url: String): String {
        val tinyResponse = client.get(TinyUrl.BASE_URL + TinyUrl.CREATE_PATH) {
            parameter("url", url)
        }

        if (!tinyResponse.status.isSuccess()) {
            error("Unable to generate TinyUrl. Response status: ${tinyResponse.status}")
        }

        val tinyUrl: String = tinyResponse.body<String>().trim()
        val tinyPath = URI(tinyUrl).path.removePrefix("/")
        check(tinyPath.isNotBlank()) { "TinyUrl response did not contain a path: $tinyUrl" }

        return URLBuilder(baseUrl)
            .apply {
                parameters.append(param, tinyPath)
            }
            .buildString()
    }

    suspend fun parse(url: URI): URI = URI(parse(url.toString()))

    suspend fun parse(url: String): String {
        val tinyUrlPath = Url(url).parameters[param]
            ?: throw IllegalArgumentException("Passed url does not contain a tiny url param")
        val followUrl = TinyUrl.BASE_URL + tinyUrlPath

        return backpedalRedirectsTillYouDie(followUrl = followUrl, expectedBaseUrl = baseUrl)
            ?: throw RedirectNotFoundException()
    }

    override fun close() {
        if (ownsClient) {
            client.close()
        }
    }

    private suspend fun backpedalRedirectsTillYouDie(
        followUrl: String,
        expectedBaseUrl: String,
        maxRedirects: Int = 20,
    ): String? {
        var nextUrl = followUrl

        repeat(maxRedirects) {
            val response = client.get(nextUrl)

            response.request.url.toString()
                .takeIf { it.contains(expectedBaseUrl) }
                ?.let { return decodeUrl(it) }

            val location = response.headers[HttpHeaders.Location]
            if (location == null) {
                return null
            }

            val resolvedUrl = resolveUrl(nextUrl, location)
            if (resolvedUrl.contains(expectedBaseUrl)) {
                return decodeUrl(resolvedUrl)
            }

            if (!response.isRedirect()) {
                return null
            }
            nextUrl = resolvedUrl
        }

        return null
    }

    private fun HttpResponse.isRedirect(): Boolean {
        return status.value in 300..399
    }

    private fun resolveUrl(currentUrl: String, location: String): String {
        return URI(currentUrl).resolve(location).toString()
    }

    private fun decodeUrl(url: String): String {
        return URLDecoder.decode(url, StandardCharsets.UTF_8)
    }

    companion object {
        private const val DEFAULT_QUERY_PARAM_TINY_URL = "tinyUrl"
    }
}
